### ISSUE TITLE
Cascade updates for Lambda

### ISSUE THREAD

**Issue by ambud** [hashicorp/terraform#8232]
Terraform version: 0.7.0
OS: Mac OS X 10.11.5 (and Ubuntu 16.04)

**Question:**
Is there a way to have dependencies in Terraform trigger a cascading update. That is if one resource changes, anything that calls it a dependency should also be updated?

Below is the template I am using, what I would like is every time the S3 Object changes, the Lambda function is recreated. The dependency is explicitly called out in this case.

```
resource "aws_s3_bucket_object" "object" {
   bucket = "${var.codebucket}"
   key = "xxxxxx.jar"
   source = "./target/xxxxxx.jar"
   etag = "${md5(file("./target/xxxxxx.jar"))}"
}

resource "aws_lambda_function" "flowlogs" {
	depends_on = ["aws_s3_bucket_object.object"]
    s3_key = "xxxxxx.jar"
    function_name = "vpcflowlogs"
    description = "xxxxxxx"
    role = "${var.lambdarole}"
    handler = "xxxxxxxxx"
    s3_bucket = "${var.codebucket}"
    runtime = "java8"
    vpc_config {
    	subnet_ids = [ "subnet-xxxxx" ]
    	security_group_ids = [ "sg-xxxxx" ]
   	}
}

resource "aws_lambda_permission" "flowlog_permission" {
  statement_id = "vpc_flow_log_activation"
  action = "lambda:InvokeFunction"
  function_name = "${aws_lambda_function.flowlogs.arn}"
  principal = "logs.us-east-1.amazonaws.com"
  source_arn = "${var.loggroup_arn}"
}

resource "aws_cloudwatch_log_subscription_filter" "flowlog_subscription_filter" {
  depends_on = ["aws_lambda_permission.flowlog_permission"]
  name = "cloudwatch_flowlog_lambda_subscription"
  log_group_name = "${var.loggroup_name}"
  filter_pattern = ""
  destination_arn = "${aws_lambda_function.flowlogs.arn}"
}
```

**Comment by apparentlymart** [comment]
Hi @ambud!

One way to do this is to make use of S3 versioning so that a new object being uploaded will cause a new version to be created. Then you can use the `s3_object_version` argument on `aws_lambda_function` to let Terraform know when it needs to upload a new version:

```hcl
resource "aws_lambda_function" "flowlogs" {
    s3_key = "${aws_s3_bucket_object.object.key}"
    s3_object_version = "${aws_s3_bucket_object.object.version_id}"
    function_name = "vpcflowlogs"
    description = "xxxxxxx"
    role = "${var.lambdarole}"
    handler = "xxxxxxxxx"
    s3_bucket = "${aws_s3_bucket_object.object.bucket}"
    runtime = "java8"
    vpc_config {
        subnet_ids = [ "subnet-xxxxx" ]
        security_group_ids = [ "sg-xxxxx" ]
    }
}
```

I also rewrote it to use interpolations from the `aws_s3_bucket_object`, which then implies the dependency between the two resources so that `depends_on` isn't necessary.

**Comment by ambud** [comment]
Got it! Thanks a lot!

**Comment by ambud** [comment]
Just tested this, while it mitigates my need to use terraform destroy. I have to use terraform apply 2 times because it's unable to plan a change in the first run.

**Comment by apparentlymart** [comment]
Hi @ambud... I think the proper fix here would be to get the patch from #6860 finished up so that we can use `source_code_hash` for S3-deployed lambda functions.

As you noticed, Terraform is not handling this correctly because it can't tell that the `version_id` will change until after it's updated the S3 object. Sorry I didn't think of that before.

I took another look at some config I use at work to do something similar to what you're doing here, and I realize that the key difference between mine and yours is that we use a different S3 key for each new version of our code, so `s3_key` changes each time we have a new version and Terraform is able to see immediately that it needs to do a new upload. If possible I'd suggest doing something like this in the mean time, until we get #6860 resolved. If you don't have a handy versioning id in your environment you might be able to do something with [the `random_id` resource](https://www.terraform.io/docs/providers/random/r/id.html), using something that changes for each deployment as a "keeper" for your resource.

(For more information on "keepers" for the random provider resources, see [the main provider docs](https://www.terraform.io/docs/providers/random/index.html).)

**Comment by ambud** [comment]
Ack. Using a build number from CI should be easy, but I guess waiting for the fix might be a better option.
